### PR TITLE
[TPC] Implement contains method for SelectionKeysSet [HZ-2265]

### PR DESCRIPTION
**Pull request by ramizdundar**
In macos, a single fd can have multiple events after a single poll. Lack
of working `contains()` method creates problems in that case.

To give more detail please look at `updateSelectedKeys` method in
`KQueueSelectorImpl` (there is a link at the bottom):
```
if (selectedKeys.contains(ski)) {
    // first time this file descriptor has been encountered on this
    // update?
    if (me.updateCount != updateCount) {
        if (ski.channel.translateAndSetReadyOps(rOps, ski)) {
            numKeysUpdated++;
            me.updateCount = updateCount;
        }
    } else {
        // ready ops have already been set on this update
        ski.channel.translateAndUpdateReadyOps(rOps, ski);
    }
} else {
    ski.channel.translateAndSetReadyOps(rOps, ski);
    if ((ski.nioReadyOps() & ski.nioInterestOps()) != 0) {
        selectedKeys.add(ski);
        numKeysUpdated++;
        me.updateCount = updateCount;
    }
}
```

Notice that if `contains()` return true and if during this poll key is
already added to the set, `translateAndUpdateReadyOps()` is called, not
`translateAndSetReadyOps()`. The difference is `translateAndSetReadyOps()`
clears the existing `readyOps` from the key. So in that scenario, we may
lose polled events because previously `contains()` method of the
`SelectorOptimizer` always returned false. So previously, the scenario
was like this:

1. A single file descriptor has 2 events: read(1) and write(4)
2. First time the code enters the above block, `readyOps` will be 1
3. Second time it enters, we put the same object to the set 2nd time (notice that our set is basically array).
4. But `translateAndSetReadyOps()` is used so `readyOps` will be 4 for both not 5
5. We lost the read event.

To prevent this we need a working `contains()` implementation.

Since there is a similar failure for Windows, I assume a similar
scenario plays out in Windows's poll implementation.

Note that this issue is present in all java versions I have tested (8,
11, 17, current main branch of openjdk).

https://github.com/openjdk/jdk8u-dev/blob/f1c1f969a77c9841cf87b2434fa90dcf4002c6be/jdk/src/macosx/classes/sun/nio/ch/KQueueSelectorImpl.java#L130

Fixes MacOS and Windows related failures of:

- https://github.com/hazelcast/hazelcast/issues/23867
- https://github.com/hazelcast/hazelcast/issues/23823
- https://github.com/hazelcast/hazelcast/issues/24076
- https://github.com/hazelcast/hazelcast/issues/23956

### CODE REVIEW

File: hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java [hunk 1]
```diff
@@ -163,6 +164,13 @@ public boolean remove(Object o) {
 
         @Override
         public boolean contains(Object o) {
+            if (!OS.isLinux()) {
```
[the remark this answers is not in the document]
Comment (pveentjer) [reply hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java:167]: On Linux the functionality isn't called. I would just provide a correct implementation and leave out the isLinux parts (so what @arodionov said).

[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java:167]: Basically we only need this implementation to work in macos and windows. This is because:

1. We never saw this reproducable failure in linux before.
2. AFAIK, a single file descriptor can have 2 events in single poll in only non-linux os. In linux polled event is single and contains both flags (read and write is 5 for example).

For this reason, I didn't want to do this extremely frequent for loop in linux. Although most of the time it will loop only once, it will still do branch predictions etc.

TLDR; this is questionable but safe optimisation

File: hazelcast/src/main/java/com/hazelcast/internal/networking/nio/SelectorOptimizer.java
[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/networking/nio/SelectorOptimizer.java:1]: https://github.com/hazelcast/hazelcast/pull/24267/commits/45b8c6866047a371c67433f846f56b3a7782eff3

File: hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ramizdundar) [reply hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java]: After talking with @pveentjer, we decided to remove this os specific logic 90d1b76.

File: hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java [hunk 2]
```diff
@@ -163,6 +163,11 @@ public boolean remove(Object o) {
 
         @Override
         public boolean contains(Object o) {
+            for (int i = 0; i < size(); i++) {
```
Comment (pveentjer) [inline hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java:166]: Can you make use of the size field directly instead of the size() method.

Comment (ramizdundar) [reply hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java]: https://github.com/hazelcast/hazelcast/pull/24267/commits/22bf636e527c2c678d7a7b891279d48734d087d9

**Review by arodionov — APPROVED** [review]

File: hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java [hunk 3]
```diff
@@ -163,6 +163,11 @@ public boolean remove(Object o) {
 
         @Override
         public boolean contains(Object o) {
+            for (int i = 0; i < activeKeys.size; i++) {
```
Comment (pveentjer) [inline hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java:166]: Can you load the activeKeys.size into a constant and use that constant to iterate over the loop. Otherwise you are at risk of calling activeKeys.size() on every iteration of the loop unless the JIT is able to optimize it out.

Comment (pveentjer) [inline hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java:166]: Can you load the activeKeys.size into a constant and use that constant to iterate over the loop. Otherwise you are at risk of calling activeKeys.size() on every iteration of the loop unless the JIT is able to optimize it out.

And can you also store the activeKey.keys array into a local variable before the loop, to increase the chances that the JIT can optimize this code better.

Comment (ramizdundar) [reply hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpcengine/nio/SelectorOptimizer.java]: https://github.com/hazelcast/hazelcast/pull/24267/commits/395b9b77cf0c96d5f8954c42ab50a896dfda4960

**Review by pveentjer — APPROVED** [review]
Great work Ramiz.

**Comment by pveentjer** [comment]
run-lab-run

**Comment by ramizdundar** [comment]
Thank you for the reviews.
